spec/requests/products/show_spec.rb

```diff
@@ -0,0 +1,61 @@
+require "web_spec"
+
+RSpec.describe "GET /products/:id", type: :request do
+  context "when product with given id does not exist" do
+    before do
+      do_request(0)
```

Comment:
I mean that in `do_request(0)` is hard to say what `0` stands for. More readable is `do_request(id: 0)` or `do_request(product_id: 0)`. It can be either a hash or keyword argument.